### ISSUE TITLE
The willSaveWaitUntil budget cannot be configured

### ISSUE THREAD

**Issue by arcanis** [microsoft/vscode#84049]
According to the [following post](https://github.com/microsoft/vscode-eslint/issues/600#issuecomment-489601617), VSCode has a time budget during which extensions must answer the save request or they get skipped. Three strikes are allowed, and the fourth time the extension gets the boot and won't be called anymore.

While it appears a good idea in theory as it ensures that extensions are gentle, in practice this behaviour is the source of a lot of frustration in my teams:

- The time budget doesn't take the CPU load in account, so running a background build may lead to extensions suddenly getting the boot for apparently no reason.

- [Some extensions try to stay within the budget at the cost of correctness](https://github.com/microsoft/vscode-eslint/issues/154) - leading their users to press Ctrl+S multiple times, defeating the purpose of the time budget and worsening DevX.

- The user isn't even made aware of what happened when an extension is killed, leading users to believe the extension is faulty when it's actually an environmental issue.

- We cannot tweak the time budget ourselves, meaning that we're completely dependant on you fixing this issue and have to wait until then (some people are switching editors because of this).

To summarise, there are multiple issues with the time budget as it's currently implemented:

  - [ ] The CPU load isn't taken into consideration when computing the save budget.
  - [ ] There is no visual feedback when an extension hook is killed.
  - [ ] There is no way to re-enable the extension without closing the window.
  - [ ] There is no way to change the time budget to unblock ourselves.
  - [ ] There is no way for extension to signal how important is the hook to the save result.

**Comment by arcanis** [comment]
I'm trying to understand a bit how the `willSaveWaitUntil` budget is implemented - I found there are two settings that seem related:

- `editor.formatOnSaveTimeout`
- `editor.codeActionsOnSaveTimeout`

However none of those seem affect `willSaveWaitUntil`, which is implemented in [vscode-languageserver-node](https://github.com/microsoft/vscode-languageserver-node). I haven't been able to locate yet the exact place where the timeout is implement (I looked for things like setTimeout or 1500, but with no relevant match). Since you implemented it do you have an idea where I can learn more about this, @dbaeumer?

**Comment by jrieken** [comment]
Can you explain what expensive operation you are running before saving which is not formatting nor code actions?

**Comment by arcanis** [comment]
I'm not running anything specific before saving. I just happen to have a very cpu intensive webpack build running in background.

To reproduce, you can artificially clog your cpu with something like `yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit & yarn tsc --noEmit`.

**Comment by jrieken** [comment]
> I'm not running anything specific before saving.

But something is running before saving, right? Do you know what that is? You can also check the extension host logs to see how what's happening (debug level). If understand correctly: if the computer is under load or slow then a static budget is not sufficient. I still wanna understand the concrete case or is this a theoretical problem?

**Comment by arcanis** [comment]
> I still wanna understand the concrete case or is this a theoretical problem?

No it's very concrete - basically the ESLint fix-errors-on-save integration simply [stops working](https://github.com/microsoft/vscode-eslint/issues/600) ([explanation here](https://github.com/microsoft/vscode-eslint/issues/600#issuecomment-489601617)) every now and then, multiple times a day. Our users have to reload the window to get them to by applied again - and that's only because we tell them, otherwise they would have no idea and would operate without autofix, thinking VSCode has a bug.

The secondary concrete problem is that the presence of the time budget is used as a [partial justification](https://github.com/microsoft/vscode-eslint/issues/154#issuecomment-253789419) for not running multiple ESLint passes, in opposition with [what the ESLint CLI does](https://github.com/eslint/eslint/blob/61749c94bd8a2ebcdfb89e0cd48c4a029a945079/lib/linter/linter.js#L1421). Our users then have to press ^S as many times as needed to get a stable output (and it's not theoretical - they actually do it), which decreases their confidence in the editor setup.

**Comment by jrieken** [comment]
Funny coincidence as @dbaeumer and I have discussed this a few days back and ESLint is becoming a code action, e.g fix all will be source code action and therefore running on save within a configurable timelimit. @dbaeumer might be able to share when those changes are shipped in ESLint

**Comment by dbaeumer** [comment]
I new version of ESLint which will not run code fixes on save using a code action will ship this month.
